src/main/webapp/app/components/commentIcon/CommentIcon.tsx

```diff
@@ -0,0 +1,160 @@
+function CommentIcon(props: ICommentIconProps) {
+  let color: string;
+  if (props.comments.length === 0) {
+    color = 'black';
+  } else if (props.comments.some(comment => comment.resolved === 'false')) {
+    color = 'gold';
+  } else {
+    color = 'green';
+  }
+
+  const [showComments, setShowComments] = useState(false);
+  const [timeoutId, setTimeoutId] = useState(null);
+  const [commentInputValue, setCommentInputValue] = useState('');
+
+  function handleMouseEnter() {
+    clearTimeout(timeoutId);
+    setShowComments(true);
+  }
+
+  function handleMouseLeave() {
+    const id = setTimeout(() => {
+      setShowComments(false);
+    }, CLOSE_POPOVER_DURATION_MILLISECONDS);
+
+    setTimeoutId(id);
+  }
+
+  return (
+    <div style={{ display: 'inline-block' }} onMouseEnter={handleMouseEnter} onMouseLeave={handleMouseLeave}>
+      <FontAwesomeIcon id="commentIcon" icon={faComment} color={color} />
+      <Popover target="commentIcon" isOpen={showComments}>
+        <PopoverBody>
+          <ListGroup flush>
+            {props.comments.map((comment, index) => (
+              <CommentItem
+                key={comment.date + comment.email}
+                comment={comment}
+                onResolveComment={() => props.onResolveComment(index)}
+                onUnresolveComment={() => props.onUnresolveComment(index)}
+                onDeleteComment={() => props.onDeleteComment(index)}
+              />
+            ))}
+          </ListGroup>
+        </PopoverBody>
+        <Container>
+          <Row className="border-top rounded-lg bg-light">
+            <InputGroup className="ml-3 mr-3 mb-3 mt-3">
+              <Input className="rounded-0" placeholder="Add comment" onChange={event => setCommentInputValue(event.target.value)} />
+              <Button className="rounded-0" outline onClick={() => props.onCreateComment(commentInputValue)}>
+                <FontAwesomeIcon icon={'plus'} />
+              </Button>
+            </InputGroup>
+          </Row>
+        </Container>
+      </Popover>
+    </div>
+  );
+}
+
+interface ICommentItemProps {
+  comment: Comment;
+  onResolveComment: () => void;
+  onUnresolveComment: () => void;
+  onDeleteComment: () => void;
+}
+
+function CommentItem(props: ICommentItemProps) {
+  const isResolved = props.comment.resolved === 'true';
+
+  const [isCounting, setIsCounting] = useState(false);
+  const [timeRemainingUntilDelete, setTimeRemainingUntilDelete] = useState(10);
+
+  const syncedIsCounting = useRef(isCounting);
+
+  useEffect(() => {
+    let timer;
+    syncedIsCounting.current = isCounting;
+
+    if (isCounting && timeRemainingUntilDelete > 0) {
+      timer = setTimeout(() => {
+        setTimeRemainingUntilDelete(time => time - 1);
+      }, 1000);
+    } else if (isCounting && timeRemainingUntilDelete === 0) {
+      props.onDeleteComment();
+      setIsCounting(false);
+      setTimeRemainingUntilDelete(10);
+    }
+
+    return () => {
+      clearTimeout(timer);
```

Comment:
By delete do you mean unmount? If, so the second useEffect handles when the component unmount. I just need this useEffect to make the timer count down from 10, and delete the comment when it reaches 0. Since we will be listening to firebase to changes from comments, when the list of comments changes in firebase, that should trigger a rerender, and the deleted comment should disappear